semgrep/semgrep/core_runner.py

```diff
@@ -182,16 +182,21 @@ async def _core_stderr_processor(
         if stream is None:
             raise RuntimeError("subprocess was created without a stream")
 
+        logger.debug("--- semgrep-core stderr ---")
+
         while True:
             # blocking read if buffer doesnt contain any lines or EOF
             line_bytes = await stream.readline()
 
             # readline returns empty when EOF
             if not line_bytes:
+                logger.debug("--- end semgrep-core stderr ---")
                 self._stderr = "".join(stderr_lines)
                 break
 
             line = line_bytes.decode("utf-8")
+            # log stderr ASAP
+            logger.debug(line.rstrip("\n"))
             stderr_lines.append(line)
```

Comment:
Do we still need to build this potentially huge list of `stderr_lines`?

Reply:
Probably not. But this gives the option to _extract_core_output() to do something else with those stderr.
I'm ok with removing it. Brendon knows better.

Reply:
Actually CI fails because of FAILED tests/e2e/test_performance.py::test_debug_performance - AssertionError...
Not sure why it's slower now.

Reply:
or maybe because the snapshot needs to be updated. the order of the logging may have changed.

Reply:
If we need to do something with the log later then perhaps we should be writing it to a temporary file instead of keeping it in memory.

Reply:
Reran tests to see if regression was a blip

Reply:
I suspect logger might be causing this to run synchronously or something, causing a perf regression?

Reply:
weird, the github-actions below says there's a potential speedup instead ...

Reply:
The perf test with the speed up doesn't run with `--debug` which is where this is slowing down

Reply:
tbh it might just be that logging ~10000 lines just does add 4 seconds :D

Reply:
Printing a lot of stuff to stdout/err does slowdown a program quite significantly.

Reply:
The old behavior was also printing stuff on stderr, but it was I guess faster because it was doing it all at once, whereas now we do it as soon as possible (which I prefer for debugging purpose).

Reply:
That's right, the old way there is buffering under the hood. What about writing the log to a file? That would buffer the writes, so it would be fast, and we would have logs even if something goes wrong.

Reply:
Isn't it enough to get that `last.log` file? Or is there a different between that you get from `last.log` and from `--debug`?

Reply:
the content of last.log and --debug should be the same (well last.log contains all the logs, so despite its name is actually not just the log of the last command)
But in both cases, if something weird happened between semgrep-core and semgrep, neither --debug nor last.log contained stuff logged in semgrep-core.

Reply:
Oh I see, of course. 🙂I was thinking that right now, given that we send everything to `semgrep-core`, we probably don't need Semgrep to do anything with the debugging log. Semgrep-core could just write the log directly to a file like `~/.semgrep/core.log`. And Semgrep could always read this file instead of keeping a potentially huge list of log entries in memory.